x-pack/plugins/enterprise_search/server/routes/app_search/engines.test.ts

```diff
@@ -0,0 +1,177 @@
+
+import { registerEnginesRoute } from './engines';
+import { ObjectType } from '@kbn/config-schema';
+
+jest.mock('node-fetch');
+const fetch = jest.requireActual('node-fetch');
+const { Response } = fetch;
+const fetchMock = require('node-fetch') as jest.Mocked<typeof fetch>;
+
+describe('engine routes', () => {
+  describe('GET /api/app_search/engines', () => {
+    const APP_SEARCH_HOST_CONFIG = 'http://localhost:3002';
+    const AUTH_HEADER = 'Basic 123';
+    const PAGE_INDEX = 1;
+    const TYPE = 'indexed';
+
+    let router: RouterMock;
+    const mockResponseFactory = httpServerMock.createResponseFactory();
+
+    const AppSearchAPI = {
+      shouldBeCalledWith(expectedUrl: string, expectedParams: object) {
+        return {
+          andReturnRedirect() {
+            fetchMock.mockImplementation((url: string, params: object) => {
+              expect(url).toEqual(expectedUrl);
+              expect(params).toEqual(expectedParams);
+
+              return Promise.resolve(
+                new Response('{}', {
+                  url: '/login',
+                })
+              );
+            });
+          },
+          andReturn(response: object) {
+            fetchMock.mockImplementation((url: string, params: object) => {
+              expect(url).toEqual(expectedUrl);
+              expect(params).toEqual(expectedParams);
+
+              return Promise.resolve(new Response(JSON.stringify(response)));
+            });
+          },
+        };
+      },
+    };
+
+    function expectResponseToBe200With(response: object) {
+      expect(mockResponseFactory.ok).toHaveBeenCalledWith(response);
+    }
+
+    async function callThisRoute(
+      request = {
+        headers: {
+          authorization: AUTH_HEADER,
+        },
+        query: {
+          type: TYPE,
+          pageIndex: PAGE_INDEX,
+        },
+      }
+    ) {
+      const [_, handler] = router.get.mock.calls[0];
+
+      const context = {} as jest.Mocked<RequestHandlerContext>;
+      await handler(context, httpServerMock.createKibanaRequest(request), mockResponseFactory);
+    }
+
+    function executeRouteValidation(data: { query: object }) {
+      const [config] = router.get.mock.calls[0];
+      const validate = config.validate as RouteValidatorConfig<{}, {}, {}>;
+      const query = validate.query as ObjectType;
+      query.validate(data.query);
+    }
+
+    beforeEach(() => {
+      jest.resetAllMocks();
+      router = mockRouter.create();
+      registerEnginesRoute({
+        router,
+        config: {
+          host: APP_SEARCH_HOST_CONFIG,
+        },
+      });
+    });
+
+    describe('when the underlying App Search API returns a 200', () => {
+      const appSearchEnginesResponse = { name: 'engine1' };
+
+      beforeEach(() => {
+        AppSearchAPI.shouldBeCalledWith(
+          `${APP_SEARCH_HOST_CONFIG}/as/engines/collection?type=${TYPE}&page[current]=${PAGE_INDEX}&page[size]=10`,
```

Comment:
I have to confess that I don't love abstracting out `http://localhost:3002`/type/etc. to a var. I know I'm not necessarily in the majority for this, but for tests I like to be able to see/read strings as their final output would be / a browser would read it, e.g.:

`http://localhost:3002/as/engines/collection?type=indexed&page[current]=1&page[size]=10`

It's a tad more annoying to keep up to date if we change it for whatever reason, but I think that tradeoff is well worth extra developer understanding - in my perfect ideal world that doesn't exist, a dev should be able to skim a test file and quickly understand what the code is trying to accomplish, what kind of data it's working with, and as a result how it works.

Reply:
I'm fine with that.

Reply:
Updated in 038b792ec5c239ce33cd95bc889359cddf5637b0